ugvc/mrd/bqsr_train_utils.py

```diff
@@ -0,0 +1,807 @@
+
+    # balance motifs
+    tp_df = data[data["label"] == "TP"]
+    fp_df = data[data["label"] == "FP"]
+
+    if params["balance_motifs"]:
+        balanced_tps = balance_df_according_motif(tp_df, "ref_alt_motif", 0)
+        if plotme:
+            plot_motif_balance(tp_df, "tps before balancing")
+            plot_motif_balance(balanced_tps, "tps after balacing")
+            plot_motif_balance(fp_df, "fps")
+        data = pd.concat([balanced_tps, fp_df], ignore_index=True)
+
+    data["label"][data["label"] == "TP"] = 1
+    data["label"][data["label"] == "FP"] = 0
+    data = data.astype({"label": "int16"})
+
+    return data, features, label_columns
+
+
+def save_model_and_data(xgb_classifier, X_test, y_test, X_train, y_train, params):
+
+    dump(xgb_classifier, params["model_save_path"])
+    for d, savename in ((X_test,params["X_test_save_path"]),(y_test,params["y_test_save_path"]),(X_train,params["X_train_save_path"]),(y_train,params["y_train_save_path"])):
+        d.to_parquet(savename)
+
+    with open(params["params_save_path"], "w", encoding="utf-8") as f:
+        json.dump(params, f)
+    
+    return
+
+
+def get_precision(df, label):
+    return (df["label"] == label).sum() / df.shape[0] if df.shape[0] > 0 else np.nan
+
+
+def get_recall(df, df_before_filter, label):
+    return (
+        (df["label"] == label).sum() / (df_before_filter["label"] == label).sum()
+        if (df_before_filter["label"] == label).sum() > 0
+        else np.nan
+    )
+
+
+def get_precision_recall(
+    df, max_score, label, score="qual", cumulative=False, apply_log_trans=True
+):
+    precisions = []
+    recalls = []
+    fprs = []
+    for i in range(max_score):
+        if cumulative:
+            d_q = df[df[score].apply(np.floor) >= i]
+        else:
+            d_q = df[df[score].apply(np.floor) == i]
+        precisions.append(get_precision(d_q, label))
+        recalls.append(get_recall(d_q, df, label))
+        if precisions[-1] == np.nan:
+            fprs.append(np.none)
+        elif apply_log_trans:
+            qual = -10 * np.log10(1 - precisions[-1])
+            qual = min(qual, max_score)
+            fprs.append(qual)
+        else:
+            fprs.append(precisions[-1])
+    return fprs, recalls
+
+
+def plot_precision_recall(lists, labels, max_score, log_scale=False):
+    for lst, label in zip(lists, labels):
+        plt.plot(lst[0:max_score], ".-", label=label)
+        plt.legend()
+        if log_scale:
+            plt.yscale("log")
+        plt.xlabel("QUAL")
+
+
+def bqsr_train(
+    hom_snv_featuremap,
+    single_substitution_featuremap,
+    out_path: str,
+    out_basename: str = "",
+    reference_fasta: str = None,
+    flow_order: str = DEFAULT_FLOW_ORDER,
+    training_chrom: str = "chr20",
+):
+    # init
+    os.makedirs(out_path, exist_ok=True)
+    if len(out_basename) > 0 and not out_basename.endswith("."):
+        out_basename += "."
+
+    params = {}    
```

Comment:
I think this params dictionary structure is bad design. It does not allow for proper interpretation of the code by the IDE, and it contains different types of arguments. You have a combination of paths to save to, input files, and parameters for the model.

I suggest the following changes:
1. The input files like hom_snv_featuremap and single_substitution_featuremap should be their own parameters
2. Arguments to the XGBoost model should be a "**xgboost_args" dictionary that you can then pass as-is to the model
xgb_classifier = xgb.XGBClassifier(**xgboost_args)
Right now you have a combination of input parameters and harcoded values. You can have hard coded defaults for convenience.
3. Consider refactoring this into a class like Doron does